File: src/main/java/com/rogeriofrsouza/app/boardgame/Piece.java
package com.rogeriofrsouza.app.boardgame;

import lombok.EqualsAndHashCode;
import lombok.Getter;

@Getter
@EqualsAndHashCode
public abstract class Piece {

    protected Position position;
    private final Board board;

    protected Piece(Board board) {
        this.board = board;
    }

    public abstract boolean[][] computePossibleMoves();

    public boolean isTargetPossibleMove(Position target) {
        return computePossibleMoves()[target.getRow()][target.getColumn()];
    }

    public boolean isThereAnyPossibleMove() {
        boolean[][] possibleMoves = computePossibleMoves();

        for (boolean[] row : possibleMoves) {
            for (boolean move : row) {
                if (move) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: src/main/java/com/rogeriofrsouza/app/boardgame/Piece.kt
package com.rogeriofrsouza.app.boardgame

abstract class Piece(val board: Board) {

    var position: Position? = null

    abstract fun computePossibleMoves(): Array<BooleanArray>

    fun isTargetPossibleMove(target: Position): Boolean {
        return board.squares[target.row][target.column].isPossibleMove
    }

    fun isThereAnyPossibleMove(): Boolean {
        return board.squares.any { row -> row.any { it.isPossibleMove } }
    }
}
